proposed/2022/package-scoring-v1.md

```diff
@@ -0,0 +1,219 @@
+
+**Issue Examples:**
+
+- Missing dependency
+- Missing README
+- No example
+- Not enough public API documentation
+- Bad semver
+- Not v1+
+- No website
+- No repository
+- Unmaintained
+- No bug tracker
+- Has CVE (Critical, High, etc)
+- Deprecated license
+- Missing license
+- Non SPDX license
+- Unsafe copyright
+- Deprecated
+- Empty package
+- etc
+
+#### Popularity
+
+Popularity is the more unique of the other three categories which are more issue based. Popularity is calculated as how many projects depend on a package over the past 30 days. This is scored as a percentile up to 100% (most used) to 0% (least used).
+
+Since this is not possible with current data pipelines, we will work on providing such an experience in the future and use two proxy values such as total weekly downloads and total count of packages depending on the package to be scored up to a percentile of up to 100%.
+
+- Total Weekly Downloads
+- Number of Dependents
+
+Total Score - 100
+
+#### Quality
+
+Quality is the combination of:
+
+- Follow NuGet conventions
+  - Provides a valid README
+- Provide documentation
+  - Provides an example
+  - 20% or more of the public API has xml doc comments
+- Platform support
+  - Supports all possible modern platforms i.e. [.NET 5+/.NET Standard](https://docs.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-1-0#net-5-and-net-standard) (iOS, Android, Windows, MacOS, Linux, etc)
+- Pass static analysis
+  - Code has no errors, warnings, lints, or formatting issues.
+- Support up-to-date dependencies
+  - All of the package dependencies are supported in the latest version
+  - Package supports the latest stable .NET SDK
+
+Each of these issues would be associated an arbitrary amount of points to make a total of 100. This definition of quality is debatable and thus we will create a minimal set of quality practices.
+
+Total Score - 100
+
+#### Maintenance
+
+Maintenance is the simplest of them all as it pertains to whether the package has been updated in a reasonable timeframe. Packages that have not been updated in more than a year may be unmaintained and any problems with the package may go unaddressed.
+
+- Package updated in the last year.
+
+The score may also be shown as a decayed value over the year approaching the year mark where it goes stale.
+
+Total Score - 100
+
+#### Security
+
+Finally, Security represents the trust indicators representing the package in the current state of the supply chain risk. This includes known vulnerabilities, license risk, and any issues surrounding secure supply chain risk.
+
+- No known security vulnerabilities (Critical, High, Moderate, Low) in top-level or transitive dependencies.
+- Specifies a valid license (SPDX, embedded)
+- Not deprecated.
```

Comment:
Should packages with installation scripts get flagged for security purposes?

Reply:
I think the concept of "flagged" in this scoring conversation is very interesting. We can have "clearly good" things like having a license and "clearly bad" things like having no content or dependencies. But what about "watch out!" things? Scripts/MSBuild tasks are not necessarily bad but they should probably be noted to the consumer.

But if we do that, then we have to think about transitivity... a sneaky package author could hide a flagged attribute in a dependency package. It gets complicated fast.

Reply:
How about making scripts/msbuild tasks more visible on a package details page? Like have a tab with its text, so one could evaluate the contents without downloading a package?